src/Interpreters/InterpreterDropQuery.cpp

```diff
@@ -238,26 +238,39 @@ BlockIO InterpreterDropQuery::executeToTableImpl(const ContextPtr & context_, AS
             else
                 table->checkTableCanBeDetached();
 
-            table->flushAndShutdown();
-            TableExclusiveLockHolder table_lock;
-
-            if (database->getUUID() == UUIDHelpers::Nil)
-                table_lock = table->lockExclusively(context_->getCurrentQueryId(), context_->getSettingsRef()[Setting::lock_acquire_timeout]);
-
-            if (query.permanently)
             {
-                /// Server may fail to restart of DETACH PERMANENTLY if table has dependent ones
-                bool check_ref_deps = getContext()->getSettingsRef()[Setting::check_referential_table_dependencies];
-                bool check_loading_deps = !check_ref_deps && getContext()->getSettingsRef()[Setting::check_table_dependencies];
-                DatabaseCatalog::instance().removeDependencies(table_id, check_ref_deps, check_loading_deps, is_drop_or_detach_database);
-                /// Drop table from memory, don't touch data, metadata file renamed and will be skipped during server restart
-                database->detachTablePermanently(context_, table_id.table_name);
-            }
-            else
-            {
-                /// Drop table from memory, don't touch data and metadata
-                database->detachTable(context_, table_id.table_name);
+                TableExclusiveLockHolder table_lock;
+                if (database->getUUID() == UUIDHelpers::Nil)
+                    table_lock = table->lockExclusively(context_->getCurrentQueryId(), context_->getSettingsRef()[Setting::lock_acquire_timeout]);
+
+                if (query.permanently)
+                {
+                    /// Server may fail to restart if we DETACH PERMANENTLY (sorry for the word order)
+                    /// a table with dependencies so we need to check them first.
+                    bool check_ref_deps = getContext()->getSettingsRef()[Setting::check_referential_table_dependencies];
+                    bool check_loading_deps = !check_ref_deps && getContext()->getSettingsRef()[Setting::check_table_dependencies];
+
+                    /// Check that table can be DETACHed. Note: method throws in case if the operation is not allowed.
+                    DatabaseCatalog::instance().removeDependencies(table_id, check_ref_deps, check_loading_deps, is_drop_or_detach_database);
```

Comment:
This method mutates the state, so it won't fix the problem, let's remove this line, merge and fix the problem in the next pr.